### PR TITLE
Do not terminate if the audio engine was stopped after a `Starling` instance was created

### PR DESCRIPTION
**Pull request by agurtovoy**
Currently, Starling makes an assumption that starting the audio engine in the initializer is enough to ensure that the engine will be running for the duration of the instance's lifetime. In practice, there is no such guarantee, and there are [quite a few](https://developer.apple.com/forums/thread/31865) [situations](https://github.com/AudioKit/AudioKit/issues/1474) where the OS can [stop the engine](https://stackoverflow.com/questions/66923293/avaudioengine-audio-stops-when-switching-the-audio-output-device) [behind the scenes](https://developer.apple.com/forums/thread/658078). Attempting to play a sound when the engine is stopped results in a terminate with `uncaught exception 'com.apple.coreaudio.avfaudio', reason: 'required condition is false: _engine->IsRunning()'.`

This PR remedies the issue by adding a corresponding check and a conditional engine restart to the `play` method. 

Sometimes it's also possible that the engine was stopped in a way that prevents it from successfully starting. The code in this PR tries to recover from these "hard" failures by resetting the engine.

### CODE REVIEW

File: Starling/Starling.swift [hunk 1]
```diff
@@ -33,6 +33,7 @@ enum StarlingError: Error {
     case resourceNotFound(name: String)
     case invalidSoundIdentifier(name: String)
     case audioLoadingFailure
+    case engineIsStopped
```
Comment (agurtovoy) [inline Starling/Starling.swift:36]: I attempt to restart the engine at the beginning of the `play` method. If the restart fails, or the engine is stopped again by the time we get to `performPlaybackOnFirstAvailablePlayer`, I report the new `engineIsStopped` non-fatal error.

File: Starling/Starling.swift [hunk 2]
```diff
@@ -55,6 +56,10 @@ public class Starling {
     private var files: [String: AVAudioFile]
     private let engine = AVAudioEngine()
 
+    // MARK: - Error handling
+
+    public static var nonFatalErrorHandler: ((Error) -> Void)? = nil
```
Comment (agurtovoy) [inline Starling/Starling.swift:61]: Added an optional non-fatal error handler so we can log failures elsewhere.

File: Starling/Starling.swift [hunk 3]
```diff
@@ -64,15 +69,8 @@ public class Starling {
...
-            players.append(createNewPlayerAttachedToEngine())
-        }
-        
-        do {
-            try engine.start()
-        } catch {
-            handleNonFatalError(error)
-        }
+        createInitialPlayers()
+        startEngine()
```
Comment (agurtovoy) [inline Starling/Starling.swift:73]: Factored out inline code into the corresponding private funcs so I can reuse it when restarting/recovering from a hard failure.

File: Starling/Starling.swift [hunk 4]
```diff
@@ -100,6 +98,10 @@ public class Starling {
     // MARK: - Public API (Playback)
     
     public func play(_ sound: SoundIdentifier, allowOverlap: Bool = true) {
+        if !engine.isRunning && !startEngine() {
```
Comment (agurtovoy) [inline Starling/Starling.swift:101]: Attempt to start the engine if not running, try resetting everything if that fails.

File: Starling/Starling.swift [hunk 5]
```diff
@@ -198,9 +205,37 @@ public class Starling {
...
+        } catch {
+            handleNonFatalError(error)
+            return false
+        }
+
+        return true
+    }
+    
+    private func resetPlayersAndEngine() {
+        engine.reset()
```
Comment (agurtovoy) [inline Starling/Starling.swift:227]: https://developer.apple.com/documentation/avfaudio/avaudioengine/1389668-reset

File: Starling/Starling.swift [hunk 6]
```diff
@@ -198,9 +205,37 @@ public class Starling {
...
+        objc_sync_enter(players)
+        players = []
+        createInitialPlayers()
+        objc_sync_exit(players)
+        
+        startEngine()
```
Comment (agurtovoy) [inline Starling/Starling.swift:233]: Re-create the players and attempt to start the engine again.

**Comment by agurtovoy** [comment]
@matthewreagan Just a gentle ping on this, please let me know if there is anything I can do to make it easier for you to review.

**Comment by matthewreagan** [comment]
Thanks @agurtovoy, will merge shortly

**Comment by agurtovoy** [comment]
@matthewreagan Thanks so much!

**Pull request merged**